pandas/core/generic.py

```diff
@@ -8812,9 +8819,32 @@ def _where(
                 raise ValueError("Array conditional must be same shape as self")
             cond = self._constructor(cond, **self._construct_axes_dict())
 
+        # Needed for DataFrames with ArrayManager, see below for details
```

Comment:
is it possible to just do this *inside* the manager itself? I know this might mean a slight refactoring here, but it would be better if we could do that

Reply:
So basically this code does two steps for `cond`: 1) align and 2) fillna. 
The way that I wrote this now, those are a bit intertwined (I check something about the dataframe that is need for fillna before the alignment step), but I don't think we want to push the alignment inside the manager? (that's something that now is always done outside I think, the manager doesn't need to care about that)

As mentioned below, I could call `cond._mgr.fillna(..)` instead of `cond.fillna()`, and then add a keyword whether to force it to be boolean or not (what is now encoded in `all_bool_columns`). 
But, in that idea, the code to determine `all_bool_columns` would still live here, and thus would only move part of the added complexity into the internals.